File: HospitalManagementSystem.Application/Services/ChatService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using HospitalManagementSystem.Application.IServices;
using HospitalManagementSystem.Domain.IRepository;
using HospitalManagementSystem.Domain.Models.Chat;

namespace HospitalManagementSystem.Application.Services
{
    public class ChatService : IChatService
    {
        private readonly IChatRepository _chatRepository;

        public ChatService(IChatRepository chatRepository)
        {
            _chatRepository = chatRepository;
        }

        // ==================== CHAT SESSIONS ====================

        public async Task<ChatSessionDto?> CreateSessionAsync(Guid patientId, Guid? doctorId, Guid? adminId, string sessionType)
        {
            var session = new ChatSession
            {
                PatientId = patientId,
                DoctorId = doctorId,
                AdminId = adminId,
                SessionType = sessionType,
                Status = "Active",
                StartedAt = DateTime.UtcNow,
                LastMessageAt = DateTime.UtcNow
            };

            var created = await _chatRepository.CreateSessionAsync(session);
            return await GetSessionByIdAsync(created.SessionId);
        }

        public async Task<ChatSessionDto?> GetSessionByIdAsync(Guid sessionId)
        {
            var session = await _chatRepository.GetSessionByIdAsync(sessionId);
            if (session == null) return null;

            var lastMessage = await _chatRepository.GetLastMessageAsync(sessionId);

            return new ChatSessionDto
            {
                SessionId = session.SessionId,
                PatientId = session.PatientId,
                PatientName = session.Patient != null ? $"{session.Patient.FirstName} {session.Patient.LastName}" : "Unknown",
                DoctorId = session.DoctorId,
                DoctorName = session.Doctor?.Name,
                AdminId = session.AdminId,
                AdminName = session.Admin?.Username,
                SessionType = session.SessionType,
                Status = session.Status,
                StartedAt = session.StartedAt,
                EndedAt = session.EndedAt,
                LastMessageAt = session.LastMessageAt,
                LastMessage = lastMessage != null ? MapMessageToDto(lastMessage) : null
            };
        }

        public async Task<List<ChatSessionDto>> GetSessionsByUserIdAsync(Guid userId, string userRole)
        {
            List<ChatSession> sessions;

            switch (userRole.ToLower())
            {
                case "doctor":
                    sessions = await _chatRepository.GetSessionsByDoctorIdAsync(userId);
                    break;
                case "admin":
                    sessions = await _chatRepository.GetSessionsByAdminIdAsync(userId);
                    break;
                default:
                    sessions = await _chatRepository.GetSessionsByPatientIdAsync(userId);
                    break;
            }

            var result = new List<ChatSessionDto>();
            foreach (var session in sessions)
            {
                var lastMessage = await _chatRepository.GetLastMessageAsync(session.SessionId);
                result.Add(new ChatSessionDto
                {
                    SessionId = session.SessionId,
                    PatientId = session.PatientId,
                    PatientName = session.Patient != null ? $"{session.Patient.FirstName} {session.Patient.LastName}" : "Unknown",
                    DoctorId = session.DoctorId,
                    DoctorName = session.Doctor?.Name,
                    AdminId = session.AdminId,
                    AdminName = session.Admin?.Username,
                    SessionType = session.SessionType,
                    Status = session.Status,
                    StartedAt = session.StartedAt,
                    EndedAt = session.EndedAt,
                    LastMessageAt = session.LastMessageAt,
                    LastMessage = lastMessage != null ? MapMessageToDto(lastMessage) : null
                });
            }

            return result;
        }

        public async Task<bool> EndSessionAsync(Guid sessionId)
        {
            var session = await _chatRepository.GetSessionByIdAsync(sessionId);
            if (session == null) return false;

            session.Status = "Ended";
            session.EndedAt = DateTime.UtcNow;
            await _chatRepository.UpdateSessionAsync(session);
            return true;
        }

        // ==================== MESSAGES ====================

        public async Task<ChatMessageDto?> SendMessageAsync(Guid sessionId, Guid senderId, string senderType, string senderName, string content, string messageType)
        {
            var message = new ChatMessage
            {
                SessionId = sessionId,
                SenderId = senderId,
                SenderType = senderType,
                SenderName = senderName,
                Content = content,
                MessageType = messageType,
                SentAt = DateTime.UtcNow
            };

            var created = await _chatRepository.CreateMessageAsync(message);
            return MapMessageToDto(created);
        }

        public async Task<List<ChatMessageDto>> GetSessionMessagesAsync(Guid sessionId, int skip = 0, int take = 50)
        {
            var messages = await _chatRepository.GetMessagesBySessionIdAsync(sessionId, skip, take);
            return messages.Select(MapMessageToDto).ToList();
        }

        public async Task<bool> MarkMessagesAsReadAsync(Guid sessionId, Guid userId)
        {
            return await _chatRepository.MarkMessagesAsReadAsync(sessionId, userId);
        }

        public async Task<int> GetUnreadCountAsync(Guid userId)
        {
            // Default to patient role for unread count
            return await _chatRepository.GetUnreadCountForUserAsync(userId, "Patient");
        }

        // ==================== CHAT REQUESTS ====================

        public async Task<ChatRequestDto?> CreateChatRequestAsync(Guid patientId, Guid doctorId, string requestType, string? message)
        {
            var request = new ChatRequest
            {
                PatientId = patientId,
                DoctorId = doctorId,
                RequestType = requestType,
                Message = message,
                Status = "Pending",
                RequestedAt = DateTime.UtcNow,
                ExpiresAt = DateTime.UtcNow.AddMinutes(5)
            };

            var created = await _chatRepository.CreateRequestAsync(request);
            var full = await _chatRepository.GetRequestByIdAsync(created.RequestId);
            return MapRequestToDto(full!);
        }

        public async Task<ChatRequestResponseDto?> AcceptChatRequestAsync(Guid requestId, Guid doctorId)
        {
            var request = await _chatRepository.GetRequestByIdAsync(requestId);
            if (request == null || request.DoctorId != doctorId || request.Status != "Pending")
                return null;

            // Create a chat session
            var session = new ChatSession
            {
                PatientId = request.PatientId,
                DoctorId = doctorId,
                SessionType = request.RequestType,
                Status = "Active",
                StartedAt = DateTime.UtcNow,
                LastMessageAt = DateTime.UtcNow
            };

            var createdSession = await _chatRepository.CreateSessionAsync(session);

            // Update request
            request.Status = "Accepted";
            request.RespondedAt = DateTime.UtcNow;
            request.SessionId = createdSession.SessionId;
            await _chatRepository.UpdateRequestAsync(request);

            // Update doctor's active chat count
            var availability = await _chatRepository.GetDoctorAvailabilityAsync(doctorId);
            if (availability != null)
            {
                availability.CurrentActiveChatCount++;
                await _chatRepository.CreateOrUpdateAvailabilityAsync(availability);
            }

            return new ChatRequestResponseDto
            {
                RequestId = requestId,
                PatientId = request.PatientId,
                SessionId = createdSession.SessionId,
                Status = "Accepted"
            };
        }

        public async Task<ChatRequestDto?> DeclineChatRequestAsync(Guid requestId, string? reason)
        {
            var request = await _chatRepository.GetRequestByIdAsync(requestId);
            if (request == null || request.Status != "Pending")
                return null;

            request.Status = "Declined";
            request.DeclineReason = reason;
            request.RespondedAt = DateTime.UtcNow;
            await _chatRepository.UpdateRequestAsync(request);

            return MapRequestToDto(request);
        }

        public async Task<ChatRequestDto?> CancelChatRequestAsync(Guid requestId)
        {
            var request = await _chatRepository.GetRequestByIdAsync(requestId);
            if (request == null || request.Status != "Pending")
                return null;

            request.Status = "Cancelled";
            request.RespondedAt = DateTime.UtcNow;
            await _chatRepository.UpdateRequestAsync(request);

            return MapRequestToDto(request);
        }

        public async Task<List<ChatRequestDto>> GetPendingRequestsForDoctorAsync(Guid doctorId)
        {
            var requests = await _chatRepository.GetPendingRequestsForDoctorAsync(doctorId);
            return requests.Select(MapRequestToDto).ToList();
        }

        public async Task<List<ChatRequestDto>> GetRequestsByPatientAsync(Guid patientId)
        {
            var requests = await _chatRepository.GetRequestsByPatientIdAsync(patientId);
            return requests.Select(MapRequestToDto).ToList();
        }

        // ==================== DOCTOR AVAILABILITY ====================

        public async Task<bool> SetDoctorAvailabilityAsync(Guid doctorId, bool isAvailableForChat, bool isAvailableForVideo, string? statusMessage, string? connectionId)
        {
            var availability = new DoctorChatAvailability
            {
                DoctorId = doctorId,
                IsAvailableForChat = isAvailableForChat,
                IsAvailableForVideo = isAvailableForVideo,
                Status = (isAvailableForChat || isAvailableForVideo) ? "Online" : "Offline",
                StatusMessage = statusMessage,
                ConnectionId = connectionId,
                LastOnlineAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            await _chatRepository.CreateOrUpdateAvailabilityAsync(availability);
            return true;
        }

        public async Task<bool> UpdateDoctorConnectionAsync(string doctorId, string connectionId)
        {
            if (!Guid.TryParse(doctorId, out var docGuid))
                return false;

            return await _chatRepository.UpdateDoctorConnectionIdAsync(docGuid, connectionId);
        }

        public async Task<bool> SetDoctorOfflineAsync(string doctorId)
        {
            if (!Guid.TryParse(doctorId, out var docGuid))
                return false;

            return await _chatRepository.SetDoctorOfflineAsync(docGuid);
        }

        public async Task<DoctorChatAvailabilityDto?> GetDoctorAvailabilityAsync(Guid doctorId)
        {
            var availability = await _chatRepository.GetDoctorAvailabilityAsync(doctorId);
            if (availability == null) return null;

            return MapAvailabilityToDto(availability);
        }

        public async Task<List<DoctorChatAvailabilityDto>> GetAvailableDoctorsAsync()
        {
            var availabilities = await _chatRepository.GetAvailableDoctorsAsync();
            return availabilities.Select(MapAvailabilityToDto).ToList();
        }

        // ==================== MAPPING HELPERS ====================

        private ChatMessageDto MapMessageToDto(ChatMessage message)
        {
            return new ChatMessageDto
            {
                MessageId = message.MessageId,
                SessionId = message.SessionId,
                SenderId = message.SenderId,
                SenderType = message.SenderType,
                SenderName = message.SenderName,
                Content = message.Content,
                MessageType = message.MessageType,
                AttachmentUrl = message.AttachmentUrl,
                AttachmentName = message.AttachmentName,
                IsRead = message.IsRead,
                ReadAt = message.ReadAt,
                SentAt = message.SentAt
            };
        }

        private ChatRequestDto MapRequestToDto(ChatRequest request)
        {
            return new ChatRequestDto
            {
                RequestId = request.RequestId,
                PatientId = request.PatientId,
                PatientName = request.Patient != null ? $"{request.Patient.FirstName} {request.Patient.LastName}" : "Unknown",
                DoctorId = request.DoctorId,
                DoctorName = request.Doctor?.Name,
                RequestType = request.RequestType,
                Status = request.Status,
                Message = request.Message,
                DeclineReason = request.DeclineReason,
                RequestedAt = request.RequestedAt,
                RespondedAt = request.RespondedAt,
                ExpiresAt = request.ExpiresAt,
                SessionId = request.SessionId
            };
        }

        private DoctorChatAvailabilityDto MapAvailabilityToDto(DoctorChatAvailability availability)
        {
            return new DoctorChatAvailabilityDto
            {
                DoctorId = availability.DoctorId,
                DoctorName = availability.Doctor?.Name ?? "Unknown",
                Specialization = availability.Doctor?.Department?.Name,
                ProfileImage = null, // Doctor model doesn't have ProfileImage
                IsAvailableForChat = availability.IsAvailableForChat,
                IsAvailableForVideo = availability.IsAvailableForVideo,
                Status = availability.Status,
                StatusMessage = availability.StatusMessage,
                LastOnlineAt = availability.LastOnlineAt,
                CurrentActiveChatCount = availability.CurrentActiveChatCount,
                MaxConcurrentChats = availability.MaxConcurrentChats
            };
        }
    }
}

File: HospitalManagementSystem-Frontend/src/core/services/chat-service.ts
import { Injectable, inject, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, Subject, BehaviorSubject } from 'rxjs';
import * as signalR from '@microsoft/signalr';
import { AccountService } from './account-service';

// ==================== INTERFACES ====================

export interface ChatSession {
  sessionId: string;
  patientId: string;
  patientName: string;
  doctorId?: string;
  doctorName?: string;
  adminId?: string;
  adminName?: string;
  sessionType: string;
  status: string;
  startedAt: string;
  endedAt?: string;
  lastMessageAt: string;
  lastMessage?: ChatMessage;
  unreadCount: number;
}

export interface ChatMessage {
  messageId: string;
  sessionId: string;
  senderId: string;
  senderType: string;
  senderName: string;
  content: string;
  messageType: string;
  attachmentUrl?: string;
  attachmentName?: string;
  isRead: boolean;
  readAt?: string;
  sentAt: string;
}

export interface ChatRequest {
  requestId: string;
  patientId: string;
  patientName: string;
  doctorId?: string;
  doctorName?: string;
  requestType: string;
  status: string;
  message?: string;
  declineReason?: string;
  requestedAt: string;
  respondedAt?: string;
  expiresAt: string;
  sessionId?: string;
}

export interface ChatRequestResponse {
  requestId: string;
  patientId: string;
  sessionId: string;
  status: string;
}

export interface DoctorChatAvailability {
  doctorId: string;
  doctorName: string;
  specialization?: string;
  profileImage?: string;
  isAvailableForChat: boolean;
  isAvailableForVideo: boolean;
  status: string;
  statusMessage?: string;
  lastOnlineAt: string;
  currentActiveChatCount: number;
  maxConcurrentChats: number;
}

export interface VideoCallInfo {
  sessionId: string;
  callerId: string;
  callerName: string;
}

export interface WebRTCSignal {
  sessionId: string;
  fromUserId: string;
  offer?: RTCSessionDescriptionInit;
  answer?: RTCSessionDescriptionInit;
  candidate?: RTCIceCandidateInit;
}

@Injectable({
  providedIn: 'root'
})
export class ChatService {
  private http = inject(HttpClient);
  private accountService = inject(AccountService);
  
  private baseUrl = 'http://localhost:5245/api/chat';
  private hubUrl = 'http://localhost:5245/hubs/chat';
  
  private hubConnection: signalR.HubConnection | null = null;
  
  // Signals for reactive state
  sessions = signal<ChatSession[]>([]);
  currentSession = signal<ChatSession | null>(null);
  messages = signal<ChatMessage[]>([]);
  pendingRequests = signal<ChatRequest[]>([]);
  availableDoctors = signal<DoctorChatAvailability[]>([]);
  unreadCount = signal<number>(0);
  isConnected = signal<boolean>(false);
  
  // Chat availability for doctors
  myChatAvailability = signal<{ isAvailableForChat: boolean; isAvailableForVideo: boolean; statusMessage?: string }>({
    isAvailableForChat: false,
    isAvailableForVideo: false
  });

  // Event subjects for real-time updates
  private messageReceived$ = new Subject<ChatMessage>();
  private chatRequestReceived$ = new Subject<ChatRequest>();
  private chatRequestAccepted$ = new Subject<ChatRequestResponse>();
  private chatRequestDeclined$ = new Subject<ChatRequest>();
  private typingUser$ = new Subject<{ userId: string; userName: string }>();
  
  // Video call events
  private incomingCall$ = new Subject<VideoCallInfo>();
  private callAccepted$ = new Subject<{ sessionId: string; acceptedBy: string }>();
  private callDeclined$ = new Subject<{ sessionId: string; declinedBy: string }>();
  private callEnded$ = new Subject<string>();
  
  // WebRTC signaling events
  private offerReceived$ = new Subject<WebRTCSignal>();
  private answerReceived$ = new Subject<WebRTCSignal>();
  private iceCandidateReceived$ = new Subject<WebRTCSignal>();

  // Public observables
  onMessageReceived = this.messageReceived$.asObservable();
  onChatRequestReceived = this.chatRequestReceived$.asObservable();
  onChatRequestAccepted = this.chatRequestAccepted$.asObservable();
  onChatRequestDeclined = this.chatRequestDeclined$.asObservable();
  onUserTyping = this.typingUser$.asObservable();
  onIncomingCall = this.incomingCall$.asObservable();
  onCallAccepted = this.callAccepted$.asObservable();
  onCallDeclined = this.callDeclined$.asObservable();
  onCallEnded = this.callEnded$.asObservable();
  onOfferReceived = this.offerReceived$.asObservable();
  onAnswerReceived = this.answerReceived$.asObservable();
  onIceCandidateReceived = this.iceCandidateReceived$.asObservable();

  // ==================== SIGNALR CONNECTION ====================

  async startConnection(): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      return;
    }

    const token = this.accountService.currentUser()?.token;
    if (!token) {
      console.error('No token available for SignalR connection');
      return;
    }

    this.hubConnection = new signalR.HubConnectionBuilder()
      .withUrl(this.hubUrl, {
        accessTokenFactory: () => token
      })
      .withAutomaticReconnect()
      .configureLogging(signalR.LogLevel.Information)
      .build();

    this.registerHubHandlers();

    try {
      await this.hubConnection.start();
      console.log('SignalR Connected');
      this.isConnected.set(true);
    } catch (err) {
      console.error('SignalR Connection Error:', err);
      this.isConnected.set(false);
    }
  }

  async stopConnection(): Promise<void> {
    if (this.hubConnection) {
      await this.hubConnection.stop();
      this.isConnected.set(false);
    }
  }

  private registerHubHandlers(): void {
    if (!this.hubConnection) return;

    // Message handlers
    this.hubConnection.on('ReceiveMessage', (message: ChatMessage) => {
      this.messages.update(msgs => [...msgs, message]);
      this.messageReceived$.next(message);
    });

    this.hubConnection.on('MessagesRead', (sessionId: string, userId: string) => {
      this.messages.update(msgs => 
        msgs.map(m => m.sessionId === sessionId && m.senderId !== userId 
          ? { ...m, isRead: true, readAt: new Date().toISOString() } 
          : m
        )
      );
    });

    // Typing indicators
    this.hubConnection.on('UserTyping', (userId: string, userName: string) => {
      this.typingUser$.next({ userId, userName });
    });

    this.hubConnection.on('UserStoppedTyping', (userId: string) => {
      this.typingUser$.next({ userId, userName: '' });
    });

    // Chat request handlers
    this.hubConnection.on('ChatRequestReceived', (request: ChatRequest) => {
      this.pendingRequests.update(reqs => [request, ...reqs]);
      this.chatRequestReceived$.next(request);
    });

    this.hubConnection.on('ChatRequestAccepted', (response: ChatRequestResponse) => {
      this.chatRequestAccepted$.next(response);
      this.loadSessions(); // Refresh sessions
    });

    this.hubConnection.on('ChatRequestDeclined', (request: ChatRequest) => {
      this.chatRequestDeclined$.next(request);
    });

    this.hubConnection.on('ChatRequestCancelled', (request: ChatRequest) => {
      this.pendingRequests.update(reqs => reqs.filter(r => r.requestId !== request.requestId));
    });

    // Doctor availability
    this.hubConnection.on('DoctorAvailabilityChanged', (availability: any) => {
      this.availableDoctors.update(docs => {
        const index = docs.findIndex(d => d.doctorId === availability.DoctorId);
        if (index >= 0) {
          docs[index] = { ...docs[index], ...availability };
          return [...docs];
        }
        return docs;
      });
    });

    // Video call handlers
    this.hubConnection.on('IncomingVideoCall', (callInfo: VideoCallInfo) => {
      this.incomingCall$.next(callInfo);
    });

    this.hubConnection.on('VideoCallAccepted', (data: { sessionId: string; acceptedBy: string }) => {
      this.callAccepted$.next(data);
    });

    this.hubConnection.on('VideoCallDeclined', (data: { sessionId: string; declinedBy: string }) => {
      this.callDeclined$.next(data);
    });

    this.hubConnection.on('VideoCallEnded', (sessionId: string) => {
      this.callEnded$.next(sessionId);
    });

    // WebRTC signaling
    this.hubConnection.on('ReceiveOffer', (signal: WebRTCSignal) => {
      this.offerReceived$.next(signal);
    });

    this.hubConnection.on('ReceiveAnswer', (signal: WebRTCSignal) => {
      this.answerReceived$.next(signal);
    });

    this.hubConnection.on('ReceiveIceCandidate', (signal: WebRTCSignal) => {
      this.iceCandidateReceived$.next(signal);
    });

    // User online/offline
    this.hubConnection.on('UserOnline', (userId: string, userRole: string) => {
      console.log(`User ${userId} (${userRole}) is online`);
    });

    this.hubConnection.on('UserOffline', (userId: string, userRole: string) => {
      console.log(`User ${userId} (${userRole}) is offline`);
    });
  }

  // ==================== HUB METHODS ====================

  async sendMessage(sessionId: string, content: string, messageType: string = 'Text'): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('SendMessage', sessionId, content, messageType);
    }
  }

  async joinSession(sessionId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('JoinSession', sessionId);
    }
  }

  async leaveSession(sessionId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('LeaveSession', sessionId);
    }
  }

  async markMessagesAsRead(sessionId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('MarkMessagesAsRead', sessionId);
    }
  }

  async startTyping(sessionId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('StartTyping', sessionId);
    }
  }

  async stopTyping(sessionId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('StopTyping', sessionId);
    }
  }

  // Chat requests via SignalR
  async requestChat(doctorId: string, requestType: string, message?: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('RequestChat', doctorId, requestType, message);
    }
  }

  async acceptChatRequest(requestId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('AcceptChatRequest', requestId);
    }
  }

  async declineChatRequest(requestId: string, reason?: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('DeclineChatRequest', requestId, reason);
    }
  }

  async cancelChatRequest(requestId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('CancelChatRequest', requestId);
    }
  }

  // Doctor availability via SignalR
  async setChatAvailability(isAvailableForChat: boolean, isAvailableForVideo: boolean, statusMessage?: string): Promise<void> {
    const user = this.accountService.currentUser();
    if (user?.role !== 'Doctor') {
      console.warn('Only doctors can set chat availability');
      return;
    }

    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('SetChatAvailability', isAvailableForChat, isAvailableForVideo, statusMessage);
      this.myChatAvailability.set({ isAvailableForChat, isAvailableForVideo, statusMessage });
    }
  }

  // Video call methods
  async initiateVideoCall(sessionId: string, targetUserId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('InitiateVideoCall', sessionId, targetUserId);
    }
  }

  async acceptVideoCall(sessionId: string, callerId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('AcceptVideoCall', sessionId, callerId);
    }
  }

  async declineVideoCall(sessionId: string, callerId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('DeclineVideoCall', sessionId, callerId);
    }
  }

  async endVideoCall(sessionId: string): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('EndVideoCall', sessionId);
    }
  }

  // WebRTC signaling
  async sendOffer(sessionId: string, targetUserId: string, offer: RTCSessionDescriptionInit): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('SendOffer', sessionId, targetUserId, offer);
    }
  }

  async sendAnswer(sessionId: string, targetUserId: string, answer: RTCSessionDescriptionInit): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('SendAnswer', sessionId, targetUserId, answer);
    }
  }

  async sendIceCandidate(sessionId: string, targetUserId: string, candidate: RTCIceCandidateInit): Promise<void> {
    if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
      await this.hubConnection.invoke('SendIceCandidate', sessionId, targetUserId, candidate);
    }
  }

  // ==================== HTTP API METHODS ====================

  loadSessions(): void {
    this.http.get<ChatSession[]>(`${this.baseUrl}/sessions`).subscribe({
      next: (sessions) => this.sessions.set(sessions),
      error: (err) => console.error('Failed to load sessions:', err)
    });
  }

  getSession(sessionId: string): Observable<ChatSession> {
    return this.http.get<ChatSession>(`${this.baseUrl}/sessions/${sessionId}`);
  }

  endSession(sessionId: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/sessions/${sessionId}/end`, {});
  }

  getMessages(sessionId: string, skip: number = 0, take: number = 50): Observable<ChatMessage[]> {
    return this.http.get<ChatMessage[]>(`${this.baseUrl}/sessions/${sessionId}/messages?skip=${skip}&take=${take}`);
  }

  loadMessages(sessionId: string): void {
    this.getMessages(sessionId).subscribe({
      next: (messages) => this.messages.set(messages),
      error: (err) => console.error('Failed to load messages:', err)
    });
  }

  getUnreadCount(): Observable<{ unreadCount: number }> {
    return this.http.get<{ unreadCount: number }>(`${this.baseUrl}/unread-count`);
  }

  loadUnreadCount(): void {
    this.getUnreadCount().subscribe({
      next: (result) => this.unreadCount.set(result.unreadCount),
      error: (err) => console.error('Failed to load unread count:', err)
    });
  }

  // Chat requests via HTTP
  createChatRequestHttp(doctorId: string, requestType: string, message?: string): Observable<ChatRequest> {
    return this.http.post<ChatRequest>(`${this.baseUrl}/requests`, { doctorId, requestType, message });
  }

  getPendingRequests(): Observable<ChatRequest[]> {
    return this.http.get<ChatRequest[]>(`${this.baseUrl}/requests/pending`);
  }

  loadPendingRequests(): void {
    this.getPendingRequests().subscribe({
      next: (requests) => this.pendingRequests.set(requests),
      error: (err) => console.error('Failed to load pending requests:', err)
    });
  }

  getMyRequests(): Observable<ChatRequest[]> {
    return this.http.get<ChatRequest[]>(`${this.baseUrl}/requests/my`);
  }

  acceptChatRequestHttp(requestId: string): Observable<ChatRequestResponse> {
    return this.http.post<ChatRequestResponse>(`${this.baseUrl}/requests/${requestId}/accept`, {});
  }

  declineChatRequestHttp(requestId: string, reason?: string): Observable<ChatRequest> {
    return this.http.post<ChatRequest>(`${this.baseUrl}/requests/${requestId}/decline`, { reason });
  }

  cancelChatRequestHttp(requestId: string): Observable<ChatRequest> {
    return this.http.post<ChatRequest>(`${this.baseUrl}/requests/${requestId}/cancel`, {});
  }

  // Doctor availability via HTTP
  getAvailableDoctors(): Observable<DoctorChatAvailability[]> {
    return this.http.get<DoctorChatAvailability[]>(`${this.baseUrl}/available-doctors`);
  }

  getAllDoctors(): Observable<DoctorChatAvailability[]> {
    return this.http.get<DoctorChatAvailability[]>(`${this.baseUrl}/doctors`);
  }

  loadAvailableDoctors(): void {
    this.getAllDoctors().subscribe({
      next: (doctors) => this.availableDoctors.set(doctors),
      error: (err) => console.error('Failed to load doctors:', err)
    });
  }

  createDirectSession(doctorId: string): Observable<ChatSession> {
    return this.http.post<ChatSession>(`${this.baseUrl}/sessions/direct`, { doctorId });
  }

  getDoctorAvailability(doctorId: string): Observable<DoctorChatAvailability> {
    return this.http.get<DoctorChatAvailability>(`${this.baseUrl}/availability/${doctorId}`);
  }

  setMyAvailabilityHttp(isAvailableForChat: boolean, isAvailableForVideo: boolean, statusMessage?: string): Observable<{ success: boolean }> {
    return this.http.post<{ success: boolean }>(`${this.baseUrl}/availability`, {
      isAvailableForChat,
      isAvailableForVideo,
      statusMessage
    });
  }
}
